File: detector.py
import re
import base58 as _base58

EVM_PATTERN = re.compile(r'\b(0x[a-fA-F0-9]{40})\b')
SOL_PATTERN = re.compile(r'\b([1-9A-HJ-NP-Za-km-z]{32,44})\b')

# Well-known Solana program addresses to ignore
_SOL_NOISE = {
    "So11111111111111111111111111111111111111112",   # wrapped SOL
    "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA",  # SPL token program
    "11111111111111111111111111111111",              # system program
    "ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJe8bv", # associated token program
    "metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s", # metaplex metadata
}


def _is_solana_mint(candidate: str) -> bool:
    # Real Solana public keys encode to exactly 32 bytes and are 43-44 base58 chars
    if not (43 <= len(candidate) <= 44):
        return False
    if candidate in _SOL_NOISE:
        return False
    try:
        decoded = _base58.b58decode(candidate)
        if len(decoded) != 32:
            return False
    except Exception:
        return False
    # Entropy guard — real addresses don't repeat a handful of chars
    if len(set(candidate)) < 10:
        return False
    return True


def extract_cas(text: str) -> list[tuple[str, str]]:
    """Return list of (address, chain) where chain is 'EVM' or 'SOL'."""
    results: list[tuple[str, str]] = []

    for addr in EVM_PATTERN.findall(text):
        results.append((addr, "EVM"))

    # Scrub EVM matches before SOL scan to avoid hex overlap
    cleaned = EVM_PATTERN.sub(" ", text)
    for candidate in SOL_PATTERN.findall(cleaned):
        if _is_solana_mint(candidate):
            results.append((candidate, "SOL"))

    return results

File: whatsapp_monitor/detector.js
// Mirrors detector.py — same EVM + Solana CA extraction logic

const EVM_PATTERN = /\b(0x[a-fA-F0-9]{40})\b/g;
const SOL_PATTERN = /\b([1-9A-HJ-NP-Za-km-z]{32,44})\b/g;

const SOL_NOISE = new Set([
  'So11111111111111111111111111111111111111112',
  'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA',
  '11111111111111111111111111111111',
  'ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJe8bv',
  'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s',
]);

const BASE58_RE = /^[1-9A-HJ-NP-Za-km-z]+$/;

function isSolanaMint(candidate) {
  if (candidate.length < 43 || candidate.length > 44) return false;
  if (SOL_NOISE.has(candidate)) return false;
  if (new Set(candidate).size < 10) return false;
  if (!BASE58_RE.test(candidate)) return false;
  return true;
}

export function extractCAs(text) {
  const results = [];

  // EVM addresses
  for (const m of text.matchAll(EVM_PATTERN)) {
    results.push({ ca: m[1], chain: 'EVM' });
  }

  // Solana — scan cleaned text so hex chars don't collide with EVM matches
  const cleaned = text.replace(EVM_PATTERN, ' ');
  for (const m of cleaned.matchAll(SOL_PATTERN)) {
    if (isSolanaMint(m[1])) {
      results.push({ ca: m[1], chain: 'SOL' });
    }
  }

  return results;
}
